cardano-api/src/Cardano/Api/DeserialiseAnyOf.hs

```diff
@@ -0,0 +1,272 @@
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE GADTs #-}
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# LANGUAGE UndecidableInstances #-}
+
+-- | Deserialisation of input that may be formatted/encoded in one of several specified
+-- ways.
+module Cardano.Api.DeserialiseAnyOf
+  ( InputFormat (..)
+  , InputDecodeError (..)
+  , deserialiseInput
+  , deserialiseInputAnyOf
+  , renderInputDecodeError
+
+  , readAndDeserialiseFromFile
+  , readAndDeserialiseFromFileAnyOf
+  , readAndDeserialiseFromTextEnvelopeFile
+  ) where
+
+import           Cardano.Prelude
+
+import           Control.Monad.Trans.Except.Extra (firstExceptT, handleIOExceptT, hoistEither)
+import qualified Data.Aeson as Aeson
+import qualified Data.ByteString as BS
+import qualified Data.ByteString.Char8 as BSC
+import qualified Data.List.NonEmpty as NE
+import qualified Data.Text as Text
+import qualified Data.Text.Encoding as Text
+
+import           Cardano.Api.Typed
+
+------------------------------------------------------------------------------
+-- Formatted/encoded input deserialisation
+------------------------------------------------------------------------------
+
+-- | Input format/encoding.
+data InputFormat a where
+  -- | Bech32 encoding.
+  InputFormatBech32 :: SerialiseAsBech32 a => InputFormat a
+
+  -- | Hex/Base16 encoding.
+  InputFormatHex :: SerialiseAsRawBytes a => InputFormat a
+
+  -- | Text envelope format.
+  InputFormatTextEnvelope :: HasTextEnvelope a => InputFormat a
+
+-- | Input decoding error.
+data InputDecodeError
+  = InputTextEnvelopeError !TextEnvelopeError
+  -- ^ The provided data seems to be a valid text envelope, but some error
+  -- occurred in deserialising it.
+  | InputBech32DecodeError !Bech32DecodeError
+  -- ^ The provided data is valid Bech32, but some error occurred in
+  -- deserialising it.
+  | InputInvalidError
+  -- ^ The provided data does not represent a valid value of the provided
+  -- type.
+  deriving (Eq, Show)
+
+instance Error InputDecodeError where
+  displayError = Text.unpack . renderInputDecodeError
+
+-- | Render an error message for a 'InputDecodeError'.
+renderInputDecodeError :: InputDecodeError -> Text
+renderInputDecodeError err =
```

Comment:
Currently, the API appears to use `render` functions as a stylistic choice.  Retaining.  If we want to change this, it can be a separate refactoring PR.